credentials/tls/certprovider/provider.go

```diff
@@ -0,0 +1,106 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package certprovider defines APIs for certificate providers in gRPC.
+//
+// Experimental
+//
+// Notice: All APIs in this package are experimental and may be removed in a
+// later release.
+package certprovider
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"strings"
+)
+
+var (
+	// ErrProviderClosed may be returned from a call to KeyMaterial() when the
+	// underlying provider instance is closed.
+	ErrProviderClosed = errors.New("provider instance is closed")
+
+	// m is a map from name to provider builder.
+	m = make(map[string]Builder)
+)
+
+// Register registers the provider builder, whose name as returned by its
+// Name() method will be used as the name registered with this builder.
+func Register(b Builder) {
+	m[strings.ToLower(b.Name())] = b
+}
+
+// Get returns the provider builder registered with the given name.
+// If no builder is registered with the provided name, nil will be returned.
+func Get(name string) Builder {
+	if b, ok := m[strings.ToLower(name)]; ok {
+		return b
+	}
+	return nil
+}
+
+// Builder creates a Provider.
+type Builder interface {
+	// Build creates a new provider with the provided config.
+	Build(StableConfig) Provider
+
+	// ParseConfig converts config input in a format specific to individual
+	// implementations and returns an implementation of the StableConfig
+	// interface.
```

Comment:
Mention something like: all equivalent configurations should return a consistent StableConfig.  Here and/or on StableConfig and/or on Canonical().

Reply:
Added a line here.